src/parser/sparqlParser/SparqlQleverVisitor.cpp

```diff
@@ -774,54 +771,41 @@ vector<SparqlTripleSimpleWithGraph> Visitor::visit(
 }
 
 // ____________________________________________________________________________________
-vector<SparqlTripleSimpleWithGraph> Visitor::transformTriplesTemplate(
-    Parser::TriplesTemplateContext* ctx,
-    const SparqlTripleSimpleWithGraph::Graph& graph) {
-  auto convertTriple = [&graph](const std::array<GraphTerm, 3>& triple)
-      -> SparqlTripleSimpleWithGraph {
-    return {visitGraphTerm(triple[0]), visitGraphTerm(triple[1]),
-            visitGraphTerm(triple[2]), graph};
-  };
-
-  return ad_utility::transform(visit(ctx), convertTriple);
-}
-
-// ____________________________________________________________________________________
-vector<SparqlTripleSimpleWithGraph> Visitor::visit(Parser::QuadsContext* ctx) {
+Quads Visitor::visit(Parser::QuadsContext* ctx) {
   // The ordering of the individual triplesTemplate and quadsNotTriples is not
   // relevant and also not known.
-  auto triplesWithGraph = ad_utility::transform(
-      ctx->triplesTemplate(), [this](Parser::TriplesTemplateContext* ctx) {
-        return transformTriplesTemplate(ctx, std::monostate{});
-      });
-  ql::ranges::move(visitVector(ctx->quadsNotTriples()),
-                   std::back_inserter(triplesWithGraph));
-  return ad_utility::flatten(std::move(triplesWithGraph));
+  Quads quads;
+  quads.freeTriples_ = ad_utility::flatten(visitVector(ctx->triplesTemplate()));
+  for (auto [graph, triples] : visitVector(ctx->quadsNotTriples())) {
```

Comment:
```suggestion
  for (auto& [graph, triples] : visitVector(ctx->quadsNotTriples())) {
```
Otherwise you have deep copies.

Reply:
Is this because structured bindings don't have/work with move?

Reply:
This has nothing todo with the structured binding, even the simple
```
for (auto s : returnVecOfStrings())
```
copies the strings, even if a temporary is returned from the function.
The reason is that the range for loop simply calls `begin` which doesn't have special semantics like "give me move-iterators when I am an RvalueReference" etc.
What would work is something like
`for (auto s: std:ranges::rvalue_view(returnVecOfStrings())` but that is C++23 or something, and also not nice.
TLDR: You have to explicitly move aut of the things in a range-for loop, not even perfect forwarding via `auto&&`  and `std::forward` will work.